stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/viewmodel/channels/ChannelViewModelFactory.kt

```diff
@@ -46,6 +47,7 @@ public class ChannelViewModelFactory(
     private val memberLimit: Int = ChannelListViewModel.DEFAULT_MEMBER_LIMIT,
     private val messageLimit: Int = ChannelListViewModel.DEFAULT_MESSAGE_LIMIT,
     private val chatEventHandlerFactory: ChatEventHandlerFactory = ChatEventHandlerFactory(chatClient.clientState),
+    public val viewModelKey: String? = null,
```

Comment:
Adding a key to the `Factory` that is not used within this class but using on a different component to decide which ViewModel to be used, could be a bit confusing for the people using this factory.
Ideally, if the inputs of the factory are the same, we should only have one ViewModel.
So, instead of injecting the `ViewModelKey` in the constructor, it can be a generated property calculated by the hash of the collaborators this factory is initialized with.
What do you think?

Reply:
I did consider this and do mostly agree, however a couple things to consider:

- In most cases if the inputs to the factory are the same then yes we should be using the same viewmodel, however there is an edge case if we have two screens showing the same list, but interactions in one screen should not affect the other.
- Adding logic that sets the key by default instead of giving the client app the option will create behavior change that could implicitely break the flow in some apps (if the app is expecting the current behaviour).

If you are guys are happy with the above though then im happy to make the change.

An alternative could be to just add the key to the `ChannelsScreen` composable parameeters, but this doesnt feel like a value that the composable should care about.

Reply:
I don't see a real use-case for the first point TBH.
Keeping this key inside of the `factory` is kind of similar to keeping it "in a global scope", I mean, it is stored somewhere to be accessed by other different component, but without a real relationship.
The one that uses this key is the `ChannelsScreen`, so we should add it there instead.

Reply:
Sounds good, then if the client app wants to generate a key from the query and filters then it can.

Reply:
<details>

<summary>Here is a new patch</summary>

```
Subject: [PATCH] #5327 moved key to ChannelsScreen.kt
---
Index: stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt
--- a/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt	(revision 911865d4fa61600df62651895e9a768485117012)
+++ b/stream-chat-android-compose-sample/src/main/java/io/getstream/chat/android/compose/sample/ui/ChannelsActivity.kt	(date 1721735007052)
@@ -21,6 +21,7 @@
 import android.os.Bundle
 import androidx.activity.compose.setContent
 import androidx.activity.viewModels
+import androidx.compose.foundation.ExperimentalFoundationApi
 import androidx.compose.foundation.background
 import androidx.compose.foundation.layout.Box
 import androidx.compose.foundation.layout.Column
@@ -31,14 +32,19 @@
 import androidx.compose.foundation.layout.padding
 import androidx.compose.foundation.layout.width
 import androidx.compose.foundation.layout.wrapContentHeight
+import androidx.compose.foundation.pager.HorizontalPager
+import androidx.compose.foundation.pager.rememberPagerState
 import androidx.compose.foundation.shape.RoundedCornerShape
 import androidx.compose.material.Scaffold
+import androidx.compose.material.Tab
+import androidx.compose.material.TabRow
 import androidx.compose.material.Text
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.collectAsState
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
 import androidx.compose.runtime.setValue
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
@@ -68,17 +74,25 @@
 import io.getstream.chat.android.models.querysort.QuerySortByField
 import kotlinx.coroutines.launch
 
+@OptIn(ExperimentalFoundationApi::class)
 class ChannelsActivity : BaseConnectedActivity() {
 
-    private val factory by lazy {
+    private val defaultFactory by lazy {
         ChannelViewModelFactory(
             ChatClient.instance(),
             QuerySortByField.descByName("last_updated"),
+        )
+    }
+
+    private val lastUpdatedFactory by lazy {
+        ChannelViewModelFactory(
+            ChatClient.instance(),
+            QuerySortByField.ascByName("last_updated"),
             null,
         )
     }
 
-    private val listViewModel: ChannelListViewModel by viewModels { factory }
+    private val listViewModel: ChannelListViewModel by viewModels { lastUpdatedFactory }
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
@@ -96,21 +110,61 @@
                 autoTranslationEnabled = ChatApp.autoTranslationEnabled,
                 allowUIAutomationTest = true,
             ) {
-                ChannelsScreen(
-                    viewModelFactory = factory,
-                    title = stringResource(id = R.string.app_name),
-                    isShowingHeader = true,
-                    searchMode = SearchMode.Messages,
-                    onChannelClick = ::openMessages,
-                    onSearchMessageItemClick = ::openMessages,
-                    onBackPressed = ::finish,
-                    onHeaderAvatarClick = {
-                        listViewModel.viewModelScope.launch {
-                            ChatHelper.disconnectUser()
-                            openUserLogin()
-                        }
-                    },
-                )
+                val coroutineScope = rememberCoroutineScope()
+
+                val tabs = listOf("Default", "Last updated")
+
+                val pagerState = rememberPagerState(
+                    pageCount = { tabs.size },
+                )
+
+                Column {
+                    TabRow(
+                        selectedTabIndex = pagerState.currentPage,
+                    ) {
+                        tabs.forEachIndexed { index, title ->
+                            Tab(
+                                selected = pagerState.currentPage == index,
+                                text = {
+                                    Text(title)
+                                },
+                                onClick = {
+                                    coroutineScope.launch {
+                                        pagerState.scrollToPage(index)
+                                    }
+                                },
+                            )
+                        }
+                    }
+
+                    HorizontalPager(
+                        state = pagerState,
+                    ) { page ->
+
+                        val (key, factory) = when (page) {
+                            0 -> "default" to defaultFactory
+                            1 -> "last-updated" to lastUpdatedFactory
+                            else -> error("unknown page")
+                        }
+
+                        ChannelsScreen(
+                            viewModelFactory = factory,
+                            viewModelKey = if (UseKeys) key else null,
+                            title = stringResource(id = R.string.app_name),
+                            isShowingHeader = true,
+                            searchMode = SearchMode.Messages,
+                            onChannelClick = ::openMessages,
+                            onSearchMessageItemClick = ::openMessages,
+                            onBackPressed = ::finish,
+                            onHeaderAvatarClick = {
+                                listViewModel.viewModelScope.launch {
+                                    ChatHelper.disconnectUser()
+                                    openUserLogin()
+                                }
+                            },
+                        )
+                    }
+                }
 
 //                MyCustomUiSimplified()
 //                MyCustomUi()
@@ -271,6 +325,10 @@
     }
 
     companion object {
+
+        /*  toggle this to see fix */
+        private const val UseKeys = false
+
         fun createIntent(context: Context): Intent {
             return Intent(context, ChannelsActivity::class.java).apply {
                 flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK

```
</details>